### PR TITLE
Migrate FindDefinitionAndHoverTests and HighlightTests part of #4173 and fix HighlightTests

### PR DESCRIPTION
**Pull request by soulomoon**

### CODE REVIEW

File: ghcide/test/exe/HighlightTests.hs [hunk 1]
```diff
@@ -44,8 +43,9 @@ tests = testGroup "highlight"
             , DocumentHighlight (R 6 10 6 13) (Just DocumentHighlightKind_Read)
             , DocumentHighlight (R 7 12 7 15) (Just DocumentHighlightKind_Read)
             ]
-  , knownBrokenForGhcVersions [GHC92, GHC94, GHC96, GHC98] "Ghc9 highlights the constructor and not just this field" $
-        testSessionWait "record" $ do
+  ,
+--   knownBrokenForGhcVersions [GHC92, GHC94, GHC96, GHC98] "Ghc9 highlights the constructor and not just this field" $
```
Comment (michaelpj) [inline ghcide/test/exe/HighlightTests.hs:47]: delete

File: ghcide/test/exe/HighlightTests.hs [hunk 2]
```diff
@@ -55,7 +55,12 @@ tests = testGroup "highlight"
           ]
         highlights <- getHighlights doc (Position 3 17)
         liftIO $ highlights @?=
-          [ DocumentHighlight (R 3 17 3 23) (Just DocumentHighlightKind_Write)
+          [
+            DocumentHighlight (R 3 11 3 14) (Just DocumentHighlightKind_Write)
+          , DocumentHighlight (R 3 17 3 23) (Just DocumentHighlightKind_Read)
+          , DocumentHighlight (R 3 32 3 38) (Just DocumentHighlightKind_Read)
+          , DocumentHighlight (R 4 4 4 7) (Just DocumentHighlightKind_Read)
+          , DocumentHighlight (R 3 17 3 23) (Just DocumentHighlightKind_Write)
```
Comment (michaelpj) [inline ghcide/test/exe/HighlightTests.hs:63]: are these outputs correct?

**Review by michaelpj — APPROVED** [review]

File: ghcide/test/exe/HighlightTests.hs (hunk 2 above)
Comment (soulomoon) [reply ghcide/test/exe/HighlightTests.hs:63]: good point should we highlight the constructor if we just hover on its field?

Comment (soulomoon) [reply ghcide/test/exe/HighlightTests.hs:63]: Is it the desire behaviour? If not I think we should kept the broken test and open an issue for it instead!

**Comment by soulomoon** [comment]
This should be the relating issue for the highlight test
https://github.com/haskell/haskell-language-server/issues/2915

File: ghcide/test/exe/HighlightTests.hs (hunk 2 above)
Comment (soulomoon) [reply ghcide/test/exe/HighlightTests.hs]: I don't think this is desire, but a fix is relatively easy. I have impelmented it.

File: ghcide/test/exe/HighlightTests.hs (hunk 1 above)
Comment (soulomoon) [reply ghcide/test/exe/HighlightTests.hs]: done

**Review by fendor — APPROVED** [review]
LGTM, thanks for taking great care of the test suite!

**Comment by michaelpj** [comment]
The fix seems plausible, but also I think there is still an underlying issue, right? So it would be nice to not forget about that.

**Comment by soulomoon** [comment]
The root is the introducion of ghc generated nodes,
but filter out them is fine for our highlights.

File: ghcide/src/Development/IDE/GHC/Compat.hs [hunk 3]
```diff
@@ -471,7 +472,8 @@ isQualifiedImport ImportDecl{ideclQualified = NotQualified} = False
 isQualifiedImport ImportDecl{}                              = True
 isQualifiedImport _                                         = False
 
-
+getSourceNodeIds :: HieAST a -> Map.Map Identifier (IdentifierDetails a)
```
[the remark this answers is not in the document]
Comment (soulomoon) [reply ghcide/src/Development/IDE/GHC/Compat.hs:476]: done

**Pull request merged**